### DISCUSSION TITLE
getting a lot of asteroids - maybe 100, maybe all of them - rather than a few specific ones, without being a Pandas guru

### DISCUSSION THREAD

**Discussion by davidmikolas** [skyfielders/python-skyfield#840] [Q&A]
Goal is to post a script that answers https://space.stackexchange.com/questions/57469/positions-of-asteroids-in-heliocentric-ecliptic-x-y-z-coordinates

To start of slow, this script should give me the first 100 asteroids as an excerpt. Note, I only know python, not good with command line nor Pandas

After downloading text file from https://www.minorplanetcenter.net/ I run this:

```
fname = 'MPCORB.DAT.txt'

with open(fname, 'r') as infile:
    lines = infile.readlines()

i_start = [i for (i, line) in enumerate(lines) if '------------------' in line][0] + 1

lines = lines[i_start:]

for line in lines[:5]:
    print(line)
print('etc.')
print('len = ', len(lines))
print('')

lines = [line for line in lines if len(line.split()) >= 24]
print('keep only lines where len(line.split()) >= 24')
print('len = ', len(lines))
print('')

print('save to MPCORB.excerpt_100.DAT')
with open ('MPCORB.excerpt_100.DAT', 'w') as outfile:
    outfile.writelines(lines[:100])
```
      
The first four lines look the same as the four lines generated by running `mpc_make_excerpt.py` and then 

```python mpc_make_excerpt.py 00001 00002 00003 00004 > MPCORB.excerpt.DAT```

in the documentation , but this way I can still get a million if I want.

Then I tried to make a list of asteroid objects in the excerpt using `for row in minor_planets.items():` because I don't know how to use Pandas, but that fails:

```
import numpy as np
import matplotlib.pyplot as plt
from skyfield.api import load, Loader
from skyfield.constants import GM_SUN_Pitjeva_2005_km3_s2 as GM_SUN
from skyfield.data import mpc

load_de = Loader('~/Documents/fishing/SkyData') # avoid multiple copies of large files

ts = load.timescale()
eph = load_de('de421.bsp')

earth, sun, moon = [eph[x] for x in ('earth', 'sun', 'moon')]

with load.open('MPCORB.excerpt_100.DAT') as f:
    minor_planets = mpc.load_mpcorb_dataframe(f)

bad_orbits = minor_planets.semimajor_axis_au.isnull()

minor_planets = minor_planets[~bad_orbits]

print(minor_planets.shape[0], 'minor planets loaded')

asteroids = []
for row in minor_planets.items():  # doesn't work
    asteroid = sun + mpc.mpcorb_orbit(row, ts, GM_SUN)
    asteroids.append(asteroid)
```
The error message is:
```
Traceback (most recent call last):
  File "asteroidz_v00.py", line 26, in <module>
    asteroid = sun + mpc.mpcorb_orbit(row, ts, GM_SUN)
  File "/Users/davido/anaconda3/lib/python3.7/site-packages/skyfield/data/mpc.py", line 81, in mpcorb_orbit
    a = row.semimajor_axis_au
AttributeError: 'tuple' object has no attribute 'semimajor_axis_au'
```

**Comment by brandon-rhodes** [comment]
Whenever I need to remember how to iterate over dataframe rows, I read back through the answers to this Stack Overflow question—see if they work for you too:

https://stackoverflow.com/questions/16476924/how-to-iterate-over-rows-in-a-dataframe-in-pandas

**Comment by davidmikolas** [comment]
Yes, this works! `for _, row in minor_planets.iterrows():` 

I'm not sure it will be fast when there's a million asteroids in my "excerpt" table, there's some warning: [Though iterating row-by-row is not especially efficient since Series objects have to be created.](https://stackoverflow.com/a/10739432/3904031)  but I'll do some more testing and reading, then I'll finish up and write an answer to that Stack Exchange question and mark your reply as an answer.

Thanks!
